src/Nelmio/Alice/Parser/RuntimeCacheParser.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+/*
+ * This file is part of the Alice package.
+ *
+ * (c) Nelmio <hello@nelm.io>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Nelmio\Alice\Parser;
+
+use Nelmio\Alice\Exception\Parser\InvalidArgumentException;
+use Nelmio\Alice\ParserInterface;
+
+/**
+ * Decorates a parser to cache the result and process includes.
+ */
+final class RuntimeCacheParser implements ParserInterface
+{
+    /**
+     * @var array[] Keys are real path of cached files and the values the resulting array
+     */
+    private $cache = [];
+
+    /**
+     * @var IncludeDataMerger
+     */
+    private $includeMerger;
+
+    /**
+     * @var ParserInterface[]
+     */
+    private $parser;
+
+    public function __construct(ParserInterface $parser)
+    {
+        $this->parser = $parser;
+        $this->includeMerger = new IncludeDataMerger();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws InvalidArgumentException
+     */
+    public function parse(string $file): array
+    {
+        $cacheResult = true;
+        $realPath = realpath($file);
+
+        // If realpath() returns false, $realPath is safely casted into an integer (i.e. an array key)
+        if (isset($this->cache[$realPath])) {
+            return $this->cache[$realPath];
+        }
+
+        // $file could not be resolved. This either means the file does not exist or is not local.
+        // This doesn't mean the file is impossible to parse, hence it is passed to the decorated
+        // parser without caching the result.
+        if (false === $realPath) {
+            $cacheResult = false;
+            $realPath = $file;
+        }
+
+        $data = $this->parser->parse($realPath);
+        if (array_key_exists('include', $data)) {
+            $data = $this->processInclude($file, $data);
+        }
+
+        if ($cacheResult) {
+            $this->cache[$realPath] = $data;
+        }
+
+        return $data;
+    }
+
+    /**
+     * @param string $file File loaded
+     * @param array  $data Parse result of the loaded file
+     *
+     * @throws InvalidArgumentException
+     * 
+     * @return array
+     */
+    private function processInclude(string $file, array $data): array
```

Comment:
Should be extracted in another parser or class dedicated to process the includes. The reason being is that the way the filepath is being resolved should be customizable; cf. https://github.com/hautelook/AliceBundle/issues/203